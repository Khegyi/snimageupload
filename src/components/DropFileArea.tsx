import CloudUploadTwoTone from '@material-ui/icons/CloudUploadTwoTone'
import React, { useEffect, useState } from 'react'
import { useRepository } from '../hooks/use-repository'

export const DropFileArea: React.FunctionComponent<{
  uploadPath: string
  style?: React.CSSProperties
  uploadsetdata: () => void
  notificationControll: (onoff: boolean) => void
  setDragOver: (onoff: boolean) => void
  isDragOver: boolean
}> = props => {
  const repo = useRepository()
  const [scrollPosition, setScroll] = useState(0)

  useEffect(() => {
    window.onscroll = function() {
      setScroll(window.pageYOffset)
    }
  }, [])

  return (
    <div
      style={{
        display: 'flex',
        flexWrap: 'wrap',
        filter: props.isDragOver ? 'blur(1px)' : undefined,
        opacity: props.isDragOver ? 0.8 : 1,
        transition: 'opacity 300ms',
        ...props.style,
      }}
      onDragEnter={ev => {
        ev.stopPropagation()
        ev.preventDefault()
        props.setDragOver(true)
      }}
      onDragLeave={ev => {
        ev.stopPropagation()
        ev.preventDefault()
        props.setDragOver(false)
      }}
      onDragOver={ev => {
        ev.stopPropagation()
        ev.preventDefault()
        props.setDragOver(true)
      }}
      onDrop={async ev => {
        ev.stopPropagation()
        ev.preventDefault()
        props.setDragOver(false)
        await repo.upload.fromDropEvent({
          event: new DragEvent('drop', { dataTransfer: ev.dataTransfer }),
          binaryPropertyName: 'Binary',
          overwrite: true,
          createFolders: true,
          parentPath: props.uploadPath ? props.uploadPath : '',
          contentTypeName: 'Image',
        })
        props.uploadsetdata()
        props.notificationControll(true)
      }}>
      <div
        style={{
          width: 'auto',
          height: 'auto',
          backgroundColor: 'transparent',
          opacity: props.isDragOver ? 0.5 : 0,
          position: 'relative',
          transition: 'opacity 300ms',
        }}>
        <CloudUploadTwoTone
          style={{ width: '100%', position: 'absolute', marginTop: `${scrollPosition}px`, height: '250px' }}
        />
      </div>
      {props.children}
    </div>
  )
}
