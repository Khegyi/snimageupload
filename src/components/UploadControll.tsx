import React from 'react'
import CloudUpload from '@material-ui/icons/CloudUpload'
import { ConstantContent } from '@sensenet/client-core'
import { Fab, IconButton } from '@material-ui/core'
import CloseIcon from '@material-ui/icons/Close'
import Snackbar from '@material-ui/core/Snackbar'
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
import { useRepository } from '../hooks/use-repository'

interface UploadControllProps {
  uploadsetdata: () => void
  notificationControll: (onoff: boolean) => void
}
const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      flexGrow: 1,
      width: '100%',
      position: 'fixed',
      zIndex: 1,
    },
    input: {
      display: 'none',
    },
    button: {
      margin: theme.spacing(1),
    },
    title: {
      marginLeft: theme.spacing(2),
      flex: 1,
    },
    appBar: {
      width: '100%',
      marginRight: '0',
    },
  }),
)

export const UploadControll: React.FunctionComponent<UploadControllProps> = props => {
  const [open, setOpen] = React.useState(false)
  const repo = useRepository()
  const classes = useStyles()
  /**
   * Opens the Snackbar for succesfull upload
   */
  function handleClick() {
    setOpen(true)
  }
  /**
   * Closes the Snackbar
   * @param e any
   */
  function handleClose(_event: React.SyntheticEvent | React.MouseEvent, reason?: string) {
    if (reason === 'clickaway') {
      return
    }
    setOpen(false)
  }
  /**
   * Handle Uploaded File
   * @param e any
   */
  async function pickFile(e: any) {
    await repo.upload.fromFileList({
      binaryPropertyName: 'Binary',
      overwrite: true,
      parentPath: `${ConstantContent.PORTAL_ROOT.Path}/Content/IT/ImageLibrary`,
      file: e.target.files[0],
      contentTypeName: 'Image',
    })
    handleClick()
    props.uploadsetdata()
    props.notificationControll(true)
  }

  return (
    <div>
      <input
        accept="image/*"
        onChange={e => pickFile(e)}
        className={classes.input}
        id="contained-button-file"
        multiple
        type="file"
      />
      <label htmlFor="contained-button-file">
        <Fab variant="extended" title="Upload" component="span" className={classes.button}>
          <CloudUpload />
        </Fab>
      </label>
    </div>
  )
}
