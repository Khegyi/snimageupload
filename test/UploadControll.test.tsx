import { mount, shallow } from 'enzyme'
import React from 'react'
import { act } from 'react-dom/test-utils'
import { Snackbar } from '@material-ui/core'
import { UploadControll } from '../src/components/UploadControll'
import { RepositoryContext } from '../src/context/repository-provider'

describe('UploadControll', () => {
  const testprop = {
    uploadsetdata: jest.fn(),
  }
  it('Matches snapshot', () => {
    const wrapper = shallow(<UploadControll {...testprop} />)
    expect(wrapper).toMatchSnapshot()
  })
  it('Upload function', async () => {
    const repository = {
      upload: {
        fromFileList: jest.fn(),
      },
    }
    const wrapper = mount(
      <RepositoryContext.Provider value={repository as any}>
        <UploadControll {...testprop} />
      </RepositoryContext.Provider>,
    )
    await act(async () => {
      wrapper.find('input').simulate('change')
    })
    expect(repository.upload.fromFileList).toBeCalled()
    expect(
      wrapper
        .update()
        .find(Snackbar)
        .prop('open'),
    ).toBe(true)
    expect(testprop.uploadsetdata).toBeCalled()
  })
  it('Close SnackBar', async () => {
    const repository = {
      upload: {
        fromFileList: jest.fn(),
      },
    }
    const wrapper = mount(
      <RepositoryContext.Provider value={repository as any}>
        <UploadControll {...testprop} />
      </RepositoryContext.Provider>,
    )
    await act(async () => {
      wrapper.find('input').simulate('change')
    })
    act(() => {
      wrapper
        .update()
        .find(Snackbar)
        .prop('onClose')!({} as any, 'a')
    })
    expect(
      wrapper
        .update()
        .find(Snackbar)
        .prop('open'),
    ).toBe(false)
  })
  it('Close SnackBar on Icon', async () => {
    const repository = {
      upload: {
        fromFileList: jest.fn(),
      },
    }
    const wrapper = mount(
      <RepositoryContext.Provider value={repository as any}>
        <UploadControll {...testprop} />
      </RepositoryContext.Provider>,
    )
    await act(async () => {
      wrapper.find('input').simulate('change')
    })
    act(() => {
      wrapper
        .update()
        .find(Snackbar)
        .prop('onClose')!({} as any, 'clickaway')
    })
    expect(
      wrapper
        .update()
        .find(Snackbar)
        .prop('open'),
    ).toBe(true)
  })
})
