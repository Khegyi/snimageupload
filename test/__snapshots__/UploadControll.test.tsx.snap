// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadControll Matches snapshot 1`] = `
<div>
  <WithStyles(ForwardRef(Snackbar))
    ContentProps={
      Object {
        "aria-describedby": "message-id",
      }
    }
    action={
      Array [
        <ForwardRef(WithStyles)
          aria-label="close"
          className="makeStyles-close-6"
          color="inherit"
          onClick={[Function]}
        >
          <Memo />
        </ForwardRef(WithStyles)>,
      ]
    }
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "bottom",
      }
    }
    autoHideDuration={6000}
    message={
      <span
        id="message-id"
      >
        Successful Upload
      </span>
    }
    onClose={[Function]}
    open={false}
  />
  <input
    accept="image/*"
    className="makeStyles-input-2"
    id="contained-button-file"
    multiple={true}
    onChange={[Function]}
    type="file"
  />
  <label
    htmlFor="contained-button-file"
  >
    <WithStyles(ForwardRef(Fab))
      className="makeStyles-button-3"
      component="span"
      title="Upload"
      variant="extended"
    >
      <CloudUploadIcon />
    </WithStyles(ForwardRef(Fab))>
  </label>
</div>
`;
